Add CoreDNS support

woopstar opened the pull request:
This PR will add CoreDNS support to Kubespray based on https://github.com/coredns/deployment

It also fixes an issue where DNS can become unavailable if a pod is not answering. This is fixed by adding a secondary dns deployment. For more information about this, see a well posted comment at https://github.com/coreos/coreos-kubernetes/issues/641#issuecomment-281174806

Fixes #1676

Atoms commented on roles/kubernetes-apps/ansible/tasks/main.yml:
```diff
@@ -11,68 +11,17 @@
-    - 'deploy'
-    - 'svc'
+- name: Kubernetes Apps | KubeDNS
+  import_tasks: tasks/kubedns.yml
```
better is to use include:
include: tasks/{{ dns_mode }}.yml
when: 
  - "dns_mode != 'nonce'"
  - "dns_mode in ['kubedns', 'coredns' ]
this way you need to write this once :)

woopstar replied:
k8s-cluster.yml file says dns mode can be:

```
# Can be dnsmasq_kubedns, kubedns, manual or none
# dns_mode: kubedns
```

There is no `dnsmasq_kubedns` file.

Atoms replied:
ok, then it makes sense

riverzhang replied:
include: tasks/{{ dns_mode }}.yml
when:

"dns_mode != 'nonce'"
"dns_mode in ['kubedns', 'coredns' ]

I also agree with this writing
Otherwise ，Failing to deploy kubedns.

The 'manifests' variable is defined twice

woopstar replied:
What about now?

Atoms commented on roles/kubernetes-apps/ansible/templates/coredns-config.yml.j2:
```diff
@@ -0,0 +1,20 @@
+    .:53 {
+        errors
+        health
+        kubernetes {{ cluster_name }} in-addr.arpa ip6.arpa {
```
it should only resolve reverse ipv6 adresses ? or there is no need to add ipv4 arpa ?

woopstar replied:
Dont know yet. This is just de default config that it uses from the link, we tested it and worked perfect. ;)

Atoms replied:
ok, then it looks good.

Atoms commented:
when done pls squash commits

woopstar commented:
This now works for me with the following set in `all.yml` group vars:

```
upstream_dns_servers:
  - 9.9.9.9
  - 8.8.8.8

docker_dns_servers_strict: false
```

And the following set in `k8s-cluster.yml` group vars:

```
# Can be dnsmasq_kubedns, kubedns, coredns, manual or none
dns_mode: coredns

# Can be docker_dns, host_resolvconf or none
resolvconf_mode: host_resolvconf
```

woopstar commented:
Here is some debug information if wanted. After deployment, the `resolv.conf` inside a pod looks like this:

```
# cat /etc/resolv.conf
nameserver 10.248.0.3
search default.svc.cluster.local svc.cluster.local cluster.local
options ndots:5
```

And doing a local lookup for a service from a busybox image resolves an IP for a service:

```
# ping kube-dns.kube-system.svc.cluster.local
PING kube-dns.kube-system.svc.cluster.local (10.248.0.3): 56 data bytes
```

On the host itself, the `resolv.conf` looks like this:

```
# cat /etc/resolv.conf
domain cluster.local
search default.svc.cluster.local svc.cluster.local
nameserver 9.9.9.9
nameserver 8.8.8.8
options ndots:2
options timeout:2
options attempts:2
```

woopstar commented:
Now confirmed working with dual deployment to fix issues with timeouts. There is a great post about the issue here: https://github.com/coreos/coreos-kubernetes/issues/641#issuecomment-281174806

```
# /opt/bin/kubectl --kubeconfig admin.conf get pods --all-namespaces
NAMESPACE     NAME                                                   READY     STATUS    RESTARTS   AGE
kube-system   coredns-7dfc584b6c-45f5v                               1/1       Running   0          55s
kube-system   coredns-7dfc584b6c-jp6t5                               1/1       Running   0          55s
kube-system   coredns-secondary-854b77d84d-4qpkg                     1/1       Running   0          55s
kube-system   coredns-secondary-854b77d84d-jhl8v                     1/1       Running   0          55s
```

```
# /opt/bin/kubectl --kubeconfig admin.conf run -i --tty busybox --image=busybox --restart=Never -- sh
If you don't see a command prompt, try pressing enter.
/ #

/ # cat /etc/resolv.conf
nameserver 10.248.0.3
nameserver 10.248.0.4
search default.svc.cluster.local svc.cluster.local cluster.local
options ndots:5

/ # ping google.dk
PING google.dk (108.177.126.94): 56 data bytes
64 bytes from 108.177.126.94: seq=0 ttl=38 time=28.580 ms
```

riverzhang commented:
@woopstar @Atoms    Coredns should also be added to the kubeadm deployment #1676

woopstar commented:
@riverzhang It should already be there. Just as deployment files instead. But I will test kubeadm now.

woopstar commented:
@riverzhang Just confirmed. CoreDNS is deployed and working with kubeadm too.

woopstar commented:
Scratch that. It deploy's both kube-dns AND CoreDNS. I'll fix that right away

woopstar commented:
Latest commit confirmed cleaning up kube-dns from deployment with kubeadm enabled.

woopstar commented:
Fixes #1676

riverzhang commented:
@woopstar  Why is coredns deployed two times?

riverzhang commented:
ci check this

woopstar commented:
@riverzhang Read the post i linked in the comment in top. CoreDNS is deployed two times to fix a case where one DNS pos is not answering/down due to host problems. It will cause DNS timeouts and the host it self will fail answering. If you deploy it two times applications will retry on the secondary dns stack.
We have encountered this alot on our production setup running 3 mio users, and when we deployed Kube-DNS in a dual stack (after we read the comments I've linked) issues went away totally.

riverzhang commented on roles/kubernetes-apps/ansible/tasks/main.yml:
```diff
@@ -11,68 +11,16 @@
-    resource: "deploy"
-    state: absent
+- name: Kubernetes Apps | DNS
+  include: tasks/{{ dns_mode }}.yml
```
it shoud be include_tasks: "tasks/{{ dns_mode }}.yml"

woopstar replied:
Fixed

woopstar replied:
Reverted this change. It makes the cleanup jobs not to run.

riverzhang commented:
there is still a problem when dns_mode is dnsmasq_kubedns,  It must create kube-dns，This will forward the k8s domain to kube-dns by dnsmasq.
https://github.com/kubernetes-incubator/kubespray/blob/master/roles/dnsmasq/templates/01-kube-dns.conf.j2#L8

woopstar commented:
Fixed it. Finally understood the problem ;)

chadswen commented on docs/vars.md:
```diff
@@ -64,6 +64,7 @@ following default cluster paramters:
 * *dns_setup* - Enables dnsmasq
 * *dnsmasq_dns_server* - Cluster IP for dnsmasq (default is 10.233.0.2)
 * *skydns_server* - Cluster IP for KubeDNS (default is 10.233.0.3)
+* *skydns_server_secondary* - Secondary Cluster IP for KubeDNS (default is 10.233.0.4)
```
Concerning cluster upgrades, would there be any guarantee that something else isn't already using 10.233.0.4 in a cluster? Seems like any service IP from `10.233.0.0/18` could be potentially taken if it wasn't reserved in the original cluster deployment.

woopstar replied:
I looked over the kubespray code, and apparently there is no special way that ip adresses are locked/guaranteed.
According to the code, the dns ip address is always made using this: `skydns_server: "{{ kube_service_addresses|ipaddr('net')|ipaddr(3)|ipaddr('address') }}"` which is by default put into both `k8s-cluster.yml` and the defaults file. So I did the same with the secondary ip address.
If the 4th ip requested as here is then added to cluster-dns to the kubelet, the ip will not be served/used for services.

If you do not use CoreDNS the 4th ip is not locked / used for anything and thereby useable inside the cluster.

I hereby say that there should be no problem at all with this method of use in kubespray deployments.

chadswen commented:
@woopstar looking good, couple of thoughts

Could we get an autoscaler for CoreDNS? It would be quite similar to the kube-dns autoscaler

On dual Deployments - If possible it would be far simpler to maintain a single deployment that has autoscaled replicas. Do the concerns I mention below explain why we can't get the same effect with one of these alternatives:
 * Create the two services, but have them both target a single deployment? If the deployment is autoscaled we should always have plenty of replicas, but I guess that would create two ip pools on the same set of pods, leading to more potential for hitting dead pods consecutively.
* One Deployment, one Service, but add the single service IP twice to `--cluster-dns=`. Perhaps some DNS clients consuming /etc/resolv.conf would not work so well if the same nameserver appears twice, but I'm not sure of those conditions (specific to certain distros maybe?)

woopstar commented:
@chadswen Allow me to try to answer this topic shortly as it's very complicated.

Let me start by saying that creating the autoscaler seems unneeded. CoreDNS performance is exceptional. According to the design proposal from (Kubernetes sig team)[https://github.com/kubernetes/community/blob/master/contributors/design-proposals/network/coredns.md#performance-and-resource-load] CoreDNS performance is sky high. Running 4 pods with CoreDNS should suffice for even very big clusters. If you cluster get's bigger than this you would most likely not use kubespray to provision your cluster but rather do it all by yourself as you would have the inhouse knowledge.

As when it comes to dual deployments. I do agree with you that maintainning a single deployment is easier, but unfortunately it will not fix the issue. If you run a single deployment you still have the issue (sergei-utinsky mentions)[https://github.com/coreos/coreos-kubernetes/issues/641#issuecomment-281174806] where if 1 coredns pod somehow fails for any reason iptables would still distribute DNS queries amongs them, including the failed one. So even if you run a single deployment, and you add two service ip's you'd still have the same problem. You would only just lower the propability of hitting that pod. CoreDNS does not support readinesschecks (yet I think), so this case is even more likely.
If you run a single clusterdns ip which is default for kubedns and coredns, you would also have the issue where applications using the Unix `getaddrinfo()` function for name resolution, and a lot of applications does that, there would be no retry of dns lookups. Application developers would then have to implement retry logic in all their applications while they often rely on their OS to do name resolution.

Writing this post actually leads me to, that we should implement (with a new PR), that the two coredns deployments are forced to different hosts so they wont run on the same mitigating this problem even more.
We saw this issue on our production cluster, which has quiet a size. We fixed this issue by actually doing three deployments swith a single pod, where each deployment was forced to a single master so each dns service was on their own host. After we did that, added `options timeout 1 attempts 1 rotate` to our containers we had no DNS issues what so ever after.
The thing is, this issue actually happens if the dns pods somehow gets evicted, a host need to update, you do maintenaince on the host or anything else where you have to take the pod "offline". You would imagine how often that actually happens. And if you, like us, runs thousands of containers, this issue becomes real.

Hope all of this makes more sense now.

mattymo commented on docs/vars.md:
```diff
@@ -63,7 +63,8 @@ following default cluster paramters:
 * *dnsmasq_dns_server* - Cluster IP for dnsmasq (default is 10.233.0.2)
-* *skydns_server* - Cluster IP for KubeDNS (default is 10.233.0.3)
+* *skydns_server* - Cluster IP for DNS (default is 10.233.0.3)
+* *skydns_server_secondary* - Secondary Cluster IP for DNS (default is 10.233.0.4)
```
This extra var is totally unnecessary. It seems like you are substituting kubedns mode with coredns mode. In this case, coredns should just use this:
```
  skydns_server: "{{ kube_service_addresses|ipaddr('net')|ipaddr(3)|ipaddr('address') }}" 
```

mattymo commented on roles/kubernetes-apps/ansible/tasks/cleanup_dns.yml:
```diff
@@ -0,0 +1,54 @@
+    - kubeadm_init.changed|default(false)
+    - inventory_hostname == groups['kube-master'][0]
+
+- name: Kubernetes Apps | Delete old KubeDNS resources
```
This should only happen when necessary. Right now it will delete kubedns always on every run, which is not what we want. So you should add a when clause to delete his if the mode is not kubedns or dnsmasq_kubedns

woopstar replied:
This is the current behavior of kubespray. See [here](https://github.com/kubernetes-incubator/kubespray/blob/master/roles/kubernetes-apps/ansible/tasks/main.yml#L27) . So impemented the same behavior in this PR.

mattymo replied:
But that cleanup was done only if kubeadm was the deploy mode.

mattymo commented on roles/kubernetes-apps/ansible/templates/coredns-deployment.yml.j2:
```diff
@@ -0,0 +1,162 @@
+    addonmanager.kubernetes.io/mode: Reconcile
+    kubernetes.io/name: "CoreDNS"
+spec:
+  replicas: 2
```
number of replicas needs to be configurable. it's okay to default to two

mattymo commented on roles/kubernetes-apps/ansible/templates/coredns-deployment.yml.j2:
```diff
@@ -0,0 +1,162 @@
+      maxSurge: 10%
+  selector:
+    matchLabels:
+      k8s-app: coredns-secondary
```
what exactly is coredns-secondary?

mattymo commented on roles/kubernetes/node/templates/kubelet.kubeadm.env.j2:
```diff
@@ -53,6 +53,8 @@ KUBELET_HOSTNAME="--hostname-override={{ kube_override_hostname }}"
 {% if dns_mode == 'kubedns' %}
 {% set kubelet_args_cluster_dns %}--cluster-dns={{ skydns_server }}{% endset %}
+{% elif dns_mode == 'coredns' %}
+{% set kubelet_args_cluster_dns %}--cluster-dns={{ skydns_server }},{{ skydns_server_secondary }}{% endset %}
```
let's get rid of this secondary server

mattymo commented on roles/kubernetes/node/templates/kubelet.standard.env.j2:
```diff
@@ -44,6 +44,8 @@ KUBELET_HOSTNAME="--hostname-override={{ kube_override_hostname }}"
 {% if dns_mode == 'kubedns' %}
 {% set kubelet_args_cluster_dns %}--cluster-dns={{ skydns_server }}{% endset %}
+{% elif dns_mode == 'coredns' %}
+{% set kubelet_args_cluster_dns %}--cluster-dns={{ skydns_server }},{{ skydns_server_secondary }}{% endset %}
```
again

mattymo commented on roles/kubernetes/preinstall/tasks/resolvconf.yml:
```diff
@@ -13,6 +13,7 @@
       options ndots:{{ ndots }}
       options timeout:2
       options attempts:2
+      rotate
```
we want the first nameserver to be picked first. please get rid of this option

woopstar commented on roles/download/defaults/main.yml:
```diff
@@ -274,25 +277,31 @@ downloads:
     sha256: "{{ dnsmasq_digest_checksum|default(None) }}"
   kubedns:
-    enabled: true
+    enabled: "{{ dns_mode != 'coredns' }}"
```
Bad idea. Then it won't get fetched when mode is `dnsmasq_kubedns`

woopstar replied:
Then what about `dnsmasq_kubedns` which also needs kubedns

mattymo commented on roles/kubernetes/preinstall/tasks/set_resolv_facts.yml:
```diff
@@ -95,6 +95,8 @@
     dnsmasq_server: |-
       {%- if dns_mode == 'kubedns' and not dns_early|bool -%}
         {{ [ skydns_server ] + upstream_dns_servers|default([]) }}
+      {%- elif dns_mode == 'coredns' and not dns_early|bool -%}
```
get rid of this line and update line 96 to match both kubedns and coredns modes

mattymo commented on roles/kubespray-defaults/defaults/main.yaml:
```diff
@@ -49,6 +49,7 @@ resolvconf_mode: docker_dns
 deploy_netchecker: false
 # Ip address of the kubernetes skydns service
 skydns_server: "{{ kube_service_addresses|ipaddr('net')|ipaddr(3)|ipaddr('address') }}"
+skydns_server_secondary: "{{ kube_service_addresses|ipaddr('net')|ipaddr(4)|ipaddr('address') }}"
```
get rid of this too

woopstar replied:
Now needed when running coredns_dual dns_mode.

chadswen commented on roles/kubernetes-apps/ansible/templates/coredns-deployment-secondary.yml.j2:
```diff
@@ -0,0 +1,81 @@
+---
```
Could we use a single template for this with the `-secondary` appended conditionally by setting a var just for the second template task? That would make this much easier to keep in sync.

woopstar replied:
How would you do so? I can't figure how this should be done to be nice and simple.

chadswen replied:
Will reply inline where the changes will be needed.

chadswen commented:
> Hope all of this makes more sense now.

@woopstar I did get all of that previously from your links and prior research, but I was mostly asking for your thoughts about the alternatives I mentioned. I'll try to ask one question at a time for clarity.

If you're using ipvs kube-proxy (or user mode for that matter) Service load balancing isn't random by default like it is with iptables. In that case, could we add the same Service IP to `--cluster-dns` twice and be guaranteed to hit a different replica of a single Deployment on the second try? See my previous message for a potential concern I'm pondering with this option.

Regardless of alternatives, I do think making dual stack optional and disabled by default is a must so thank you for that.

woopstar commented:
Sorry @chadswen, I missed the part about ipvs. I think you're right about the ipvs situation though. The above concert an explination I tried to give is solely based on iptables, sinc ipvs just came and soon to be stable(?).

If you run ipvs mode for kube-proxy I do think by adding a secondary dns server that actually points to the same deployment would work out. I havn't tested it and I'm unsure howto since my expirence comes from 2 years with iptables and a big ass cluster, but the solution sounds plausible because ipvs is then route based and not iptables based.
I'm unsure how it's handled since CoreDNS does not have a readiness check and if that has any impact at all.

We could add a condition that if `kube_proxy_mode` is set to `ipvs` and you set `dns_mode` to `coredns` a secondary service is deployed that points to the same deployment, and the secondary ip used for the service is added to `--cluster-dns` to achieve this behavior?

chadswen commented:
@woopstar I think we're fine leaving dual Deployment as an off by default toggle. Keep it simple for now.

In time, if ipvs becomes the kubespray default, we could take another look at the advantages of the single Deployment + multiple `--cluster-ip` values approach.

I do agree that healthchecks are crucially important for any cluster infra Service regardless of kube-proxy mode. Hopefully that capability comes soon.

chadswen commented on roles/kubernetes-apps/ansible/tasks/coredns.yml:
```diff
@@ -0,0 +1,34 @@
+  tags:
+    - coredns
+
+- name: Kubernetes Apps | Lay Down Secondary CoreDNS Template
```
On this task, you could set a task only var `coredns_secondary_tmpl: true` or something, and target the same same template file as the above task. Something like
```yaml
- name: Kubernetes Apps | Lay Down Secondary CoreDNS Template
  template:
    src: "{{ item.file }}.j2"
    dest: "{{ kube_config_dir }}/{{ item.file }}"
  with_items:
    - { name: coredns, file: coredns-deployment.yml, type: deployment }
    - { name: coredns, file: coredns-svc.yml, type: svc }
  vars:
    coredns_ordinal: "-secondary"
...
```

woopstar replied:
Wont be possible to do so. It would overwrite the file with the first svc and deployment in it?

chadswen replied:
```yaml
- name: Kubernetes Apps | Lay Down Secondary CoreDNS Template
  template:
    src: "{{ item.src}}"
    dest: "{{ kube_config_dir }}/{{ item.file }}"
  with_items:
    - { name: coredns, file: coredns-deployment-secondary.yml, src: coredns-deployment.yml.j2, type: deployment }
    - { name: coredns, file: coredns-svc-secondary.yml, src: coredns-svc.yml.j2 , type: svc }
  vars:
    coredns_ordinal: "-secondary"
...
```

chadswen commented on roles/kubernetes-apps/ansible/templates/coredns-deployment-secondary.yml.j2:
```diff
@@ -0,0 +1,81 @@
+  name: coredns-secondary
+  namespace: {{ system_namespace }}
+  labels:
+    k8s-app: coredns-secondary
```
Here and everywhere else you append -secondary in both the deployment and service, replace that with a jinja2 filter, something like 
```yaml
  name: coredns{{ coredns_ordinal_suffix | default(omit) }}
```
With that done, you could just use the single template with dynamic names.

woopstar commented:
PR fixed with all of @chadswen's inputs. Ready for merge?
